internal/controller/pipelinerollout_progressive.go

```diff
@@ -90,20 +92,25 @@ func (r *PipelineRolloutReconciler) processUpgradingPipelineStatus(
 		}
 	}
 
-	pipelineStatus, err := kubernetes.ParseStatus(existingUpgradingPipelineDef)
+	pipelineStatus, err := parsePipelineStatus(existingUpgradingPipelineDef)
 	if err != nil {
 		return false, fmt.Errorf("failed to parse Pipeline Status from pipeline CR: %+v, %v", existingUpgradingPipelineDef, err)
 	}
 
-	pipelinePhase := numaflowv1.PipelinePhase(pipelineStatus.Phase)
+	pipelinePhase := pipelineStatus.Phase
 	if pipelinePhase == numaflowv1.PipelinePhaseFailed {
-		err = r.updatePipelineLabel(ctx, r.restConfig, existingUpgradingPipelineDef, "")
+		// Mark the failed new pipeline recyclable.
+		err = r.updatePipelineLabel(ctx, r.restConfig, existingUpgradingPipelineDef, string(common.LabelValueUpgradeRecyclable))
 		if err != nil {
 			return false, err
 		}
 		pipelineRollout.Status.MarkPipelineProgressiveUpgradeFailed("New Pipeline Failed", pipelineRollout.Generation)
 		return false, nil
 	} else if pipelinePhase == numaflowv1.PipelinePhaseRunning {
+		if !isPipelineReady(pipelineStatus.Status) {
+			//continue (re-enqueue)
+			return false, nil
+		}
```

Comment:
I know this has the ambiguity of the Child Conditions and whether they mean "Progressing" or actually unhealthy, which we can't really do anything about. So, I guess if they are actually unhealthy, then the PipelineRollout will simply remain in a Pending state with this upgrading Pipeline, and then the user can figure out how to resolve it. Does that sound right?

Reply:
that's right